util/tracer_nvbit/tracer_tool/tracer_tool.cu

```diff
@@ -348,182 +549,152 @@ void nvbit_at_cuda_event(CUcontext ctx, int is_exit, nvbit_api_cuda_t cbid,
     fclose(statsFile);
   }
 
-  if (cbid == API_CUDA_cuMemcpyHtoD_v2) {
+  switch (cbid) {
+  // We start with recording kernel launch events
+  // Identify all the possible CUDA launch events without stream
+  // parameters, they will not get involved with cuda graph
+  case API_CUDA_cuLaunch:
+  case API_CUDA_cuLaunchGrid: {
+    cuLaunch_params *p = (cuLaunch_params *)params;
+    CUfunction func = p->f;
     if (!is_exit) {
-      cuMemcpyHtoD_v2_params *p = (cuMemcpyHtoD_v2_params *)params;
-      char buffer[1024];
-      kernelsFile = fopen(ctx_kernelslist[ctx].c_str(), "a");
-      sprintf(buffer, "MemcpyHtoD,0x%016lx,%lld", p->dstDevice, p->ByteCount);
-      fprintf(kernelsFile, buffer);
-      fprintf(kernelsFile, "\n");
-      fclose(kernelsFile);
+      enter_kernel_launch(ctx, func, cbid, params, false, false);
+    } else {
+      leave_kernel_launch(ctx);
+    }
+  } break;
+  // To support kernel launched by cuda graph (in addition to existing kernel
+  // launche method), we need to do:
+  //
+  // 1. instrument kernels at cudaGraphAddKernelNode event. This is for cases
+  // that kernels are manually added to a cuda graph.
+  // 2. distinguish captured kernels when kernels are recorded to a graph
+  // using stream capture. cudaStreamIsCapturing() tells us whether a stream
+  // is capturiong.
+  // 3. per-kernel instruction counters, since cuda graph can launch multiple
+  // kernels at the same time.
+  //
+  // Three cases:
+  //
+  // 1. original kernel launch:
```

Comment:
Damn, CUDA graph is a huge pain.my understanding was that the current NVBit does not support cuda graph applications, but it looks like your code handles them now? Is there some NVBit version update we need to support this?

Reply:
Actually, I just followed nvbit's `mem_trace` tool to handle cuda graph. All of these comments come from their tool, so I think they should support cuda graph now.

Reply:
The `mem_trace` tool is already in the NVBit 1.7.4 release.